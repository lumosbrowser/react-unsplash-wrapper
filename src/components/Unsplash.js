import React from "react"
import PropTypes from "prop-types"

// Constants
const PATH = "//source.unsplash.com"
const USER = "user"
const COLLECTION = "collection"
const DAILY = "daily"
const RANDOM = "random"
const WIDTH = 2160
const HEIGHT = 400
const DEFAULT_STYLES = {
  backgroundRepeat: "no-repeat",
  backgroundSize: "cover",
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  alignItems: "center",
  margin: "auto"
}
const EXPAND_STYLES = {
  position: "absolute",
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  margin: 0
}

// Helper to generate url for unsplash
const generateUrl = ({
  username,
  width,
  height,
  photoId,
  collectionId,
  keywords,
  fixed
}) => {
  const url = [PATH]

  if (fixed) return url.push(DAILY).join("/")

  if (username) url.push(USER, username)
  if (!username && collectionId) url.push(COLLECTION, collectionId)
  if (!username && !collectionId && photoId) url.push(photoId)
  if (!username && !photoId && !collectionId && !keywords) url.push(RANDOM)

  url.push(`${width}x${height}`)

  if (keywords) url.push(`?${keywords.replace(/\s/g, "")}`)

  return url.join("/")
}

const Unsplash = ({
  children,
  photoId,
  collectionId,
  username,
  expand,
  fixed,
  img,
  keywords,
  style,
  width,
  height
}) => {
  if (typeof width === "string") width = parseInt(width, 10)
  if (typeof height === "string") height = parseInt(height, 10)

  const urlWidth = (style && style.width) || width
  const urlHeight = (style && style.height) || height

  const url = encodeURI(
    generateUrl({
      username,
      collectionId,
      photoId,
      keywords,
      fixed,
      width: urlWidth,
      height: urlHeight
    })
  )

  let imageStyles = {
    width,
    height,
    ...style
  }

  let backgroundStyles = {
    ...DEFAULT_STYLES,
    backgroundImage: `url(${url})`
  }

  if (expand) {
    backgroundStyles = { ...backgroundStyles, ...EXPAND_STYLES }
  } else {
    backgroundStyles = {
      ...backgroundStyles,
      width,
      height,
      ...style
    }
  }

  return img ? (
    <img src={url} style={imageStyles} />
  ) : (
    <div style={backgroundStyles}>{children}</div>
  )
}

Unsplash.propTypes = {
  children: PropTypes.node,
  collectionId: PropTypes.number,
  username: PropTypes.string,
  keywords: PropTypes.string,
  expand: PropTypes.bool,
  fixed: PropTypes.bool,
  img: PropTypes.bool,
  style: PropTypes.object,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}

Unsplash.defaultProps = {
  expand: false,
  fixed: false,
  img: false,
  width: WIDTH,
  height: HEIGHT
}

export default Unsplash
