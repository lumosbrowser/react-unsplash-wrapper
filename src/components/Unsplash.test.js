import React from "react"
import renderer from "react-test-renderer"
import Unsplash from "./Unsplash"

describe("Unsplash", () => {
  it("renders without crashing", () => {
    const tree = renderer.create(<Unsplash />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a random image", () => {
    const tree = renderer.create(<Unsplash img />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows an avatar", () => {
    const tree = renderer
      .create(<Unsplash width="64" height="64" keywords="face" img />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a placeholder with image, size and random image", () => {
    const tree = renderer.create(<Unsplash width="800" height="200" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a placeholder with custom size and keywords", () => {
    const tree = renderer
      .create(
        <Unsplash width="800" height="200" keywords="beach, palms, sea" />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a placeholder with content inside and custom styles", () => {
    const tree = renderer
      .create(
        <Unsplash style={{ lineHeight: 2 }}>
          <h1 style={{ color: "white", textShadow: "1px 1px 2px black" }}>
            Super awesome title
          </h1>
          <p>
            With React Unsplash Wrapper is really easy to create a Hero image.
          </p>
        </Unsplash>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a placeholder that expands to its parent", () => {
    const tree = renderer
      .create(
        <div style={{ position: "relative", width: 400, height: 400 }}>
          <Unsplash expand />
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a random image from a user", () => {
    const tree = renderer.create(<Unsplash username="erondu" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a random image from a collection", () => {
    const tree = renderer.create(<Unsplash collectionId={190727} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("shows a specific photo id", () => {
    const tree = renderer
      .create(<Unsplash photoId="WLUHO9A_xik" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
