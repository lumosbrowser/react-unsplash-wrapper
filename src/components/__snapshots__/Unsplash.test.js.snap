// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unsplash renders without crashing 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundImage": "url(//source.unsplash.com/random/1080x720)",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "display": "flex",
      "flexDirection": "column",
      "height": 720,
      "justifyContent": "center",
      "margin": "auto",
      "width": 1080,
    }
  }
/>
`;

exports[`Unsplash shows a placeholder that expands to its parent 1`] = `
<div
  style={
    Object {
      "height": 400,
      "position": "relative",
      "width": 400,
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "backgroundImage": "url(//source.unsplash.com/random/1080x720)",
        "backgroundRepeat": "no-repeat",
        "backgroundSize": "cover",
        "bottom": 0,
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "left": 0,
        "margin": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</div>
`;

exports[`Unsplash shows a placeholder with content inside and custom styles 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundImage": "url(//source.unsplash.com/random/1080x720)",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "display": "flex",
      "flexDirection": "column",
      "height": 720,
      "justifyContent": "center",
      "lineHeight": 2,
      "margin": "auto",
      "width": 1080,
    }
  }
>
  <h1
    style={
      Object {
        "color": "white",
        "textShadow": "1px 1px 2px black",
      }
    }
  >
    Super awesome title
  </h1>
  <p>
    With React Unsplash Wrapper is really easy to create a Hero image.
  </p>
</div>
`;

exports[`Unsplash shows a placeholder with custom size and keywords 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundImage": "url(//source.unsplash.com/800x200/?beach,palms,sea)",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "display": "flex",
      "flexDirection": "column",
      "height": 200,
      "justifyContent": "center",
      "margin": "auto",
      "width": 800,
    }
  }
/>
`;

exports[`Unsplash shows a placeholder with image, size and random image 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundImage": "url(//source.unsplash.com/random/800x200)",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "display": "flex",
      "flexDirection": "column",
      "height": 200,
      "justifyContent": "center",
      "margin": "auto",
      "width": 800,
    }
  }
/>
`;

exports[`Unsplash shows a random image 1`] = `
<img
  src="//source.unsplash.com/random/1080x720"
  style={
    Object {
      "height": 720,
      "width": 1080,
    }
  }
/>
`;

exports[`Unsplash shows an avatar 1`] = `
<img
  src="//source.unsplash.com/64x64/?face"
  style={
    Object {
      "height": 64,
      "width": 64,
    }
  }
/>
`;
